### PR TITLE
BAVL-98 initial first attempt at the beginnings of entity modelling for the create a booking journey.

### PR DESCRIPTION
**Pull request by matt-crowson-moj**
The work in progress for this PR now goes a stage further to start creating the video bookings in the service layer.  This is still pretty basic.

I will soon start adding basic controller and integration tests so we have some real end to end testing going on.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/entity/VideoBooking.kt [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import jakarta.persistence.Id
+import jakarta.persistence.JoinColumn
+import jakarta.persistence.OneToOne
+import jakarta.persistence.Table
+import org.hibernate.Hibernate
+import java.time.LocalDateTime
+
+@Entity
+@Table(name = "video_booking")
+class VideoBooking private constructor(
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/entity/VideoBooking.kt:15]: Note the use of private constructor.

Have companion object factory methods to support the different types i.e. court and probation.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/entity/VideoBooking.kt [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    ): VideoBooking =
+      VideoBooking(
+        bookingType = "PROBATION",
+        court = null,
+        hearingType = null,
+        probationTeam = probationTeam,
+        probationMeetingType = probationMeetingType,
+        videoUrl = videoUrl,
+        createdBy = createdBy,
+      )
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/entity/VideoBooking.kt:93]: Neat :-)

**Review by timharrison-moj — APPROVED** [review]
LGTM.
For court and probation team, we could add the list of contacts in a 1..m relationship?  What I'd like to avoid is to link the booking history to the booking ..

**Comment by matt-crowson-moj** [comment]
> LGTM. For court and probation team, we could add the list of contacts in a 1..m relationship? What I'd like to avoid is to link the booking history to the booking ..

Linking the contacts sounds good.  This shouldn't be too much data.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/BookingFacade.kt [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+@Component
+class BookingFacade(
+  private val createVideoBookingService: CreateVideoBookingService,
+  private val emailService: EmailService,
+) {
+  companion object {
+    private val log = LoggerFactory.getLogger(this::class.java)
+  }
+
+  fun create(booking: CreateVideoBookingRequest): Long =
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/BookingFacade.kt:21]: We may want to return some better model representation of the newly created booking but for now just returning the identifier.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/prisonersearch/PrisonerSearchClient.kt [hunk 4]
```diff
@@ -14,6 +17,10 @@ class PrisonerSearchClient(private val prisonerSearchApiWebClient: WebClient) {
       .bodyToMono(Prisoner::class.java)
       .onErrorResume(WebClientResponseException.NotFound::class.java) { Mono.empty() }
       .block()
+
+  fun getPrisonerAtPrison(prisonerNumber: String, prisonCode: String): Prisoner =
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/prisonersearch/PrisonerSearchClient.kt:21]: Put this in here for convenience.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/model/request/CreateVideoBookingRequest.kt [hunk 5]
```diff
@@ -39,12 +39,19 @@ data class CreateVideoBookingRequest(
   @field:Size(max = 400, message = "Comments for the video link booking cannot not exceed {max} characters")
   @Schema(description = "Free text comments for the video link booking", example = "Waiting to hear on legal representation")
   val comments: String?,
+
+  @field:Size(max = 120, message = "The video link should not exceed {max} characters")
+  @Schema(description = "The video link for the appointment. Must be a valid URL", example = "https://video.here.com")
+  val videoLinkUrl: String?,
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/model/request/CreateVideoBookingRequest.kt:45]: This was in the wrong place.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/model/request/CreateVideoBookingRequest.kt [hunk 6]
```diff
@@ -124,14 +135,18 @@ data class Appointment(
...
 }
+
+enum class AppointmentType {
+  // Probation types
+  RECALL_REPORT,
+  PRE_SENTENCE_REPORT,
+
+  // Court types
+  PRE_CONFERENCE,
+  HEARING,
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/model/request/CreateVideoBookingRequest.kt:150]: I will mirror these descriptions in the reference_codes data too - was just going to expand the court types into PRE, MAIN and POST.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/CreateVideoBookingService.kt [hunk 7]
```diff
@@ -1,11 +1,75 @@
...
+      court = court,
+      hearingType = request.courtHearingType!!.name,
+      videoUrl = request.videoLinkUrl,
+      createdBy = "TBD",
+    )
+
+    log.info("TODO - create the appointments linked to the booking.")
+
+    val persistedBooking = videoBookingRepository.save(newCourtBooking)
+
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/CreateVideoBookingService.kt:49]: Will probably add the appointment creation in here ...

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/CreateVideoBookingService.kt [hunk 8]
```diff
@@ -1,11 +1,75 @@
...
+      probationTeam = probationTeam,
+      probationMeetingType = request.probationMeetingType!!.name,
+      videoUrl = request.videoLinkUrl,
+      createdBy = "TBD",
+    )
+
+    log.info("TODO - create the appointments linked to the booking.")
+
+    val persistedBooking = videoBookingRepository.save(newProbationTeamBooking)
+
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/CreateVideoBookingService.kt:69]: Will probably add the appointment creation in here ...

**Review by timharrison-moj — APPROVED** [review]
LGTM

**Pull request merged**